packages/@aws-cdk/aws-lambda/lib/function.ts

```diff
@@ -824,17 +824,7 @@ export class Function extends FunctionBase {
       }
     }
 
-    // Pick subnets, make sure they're not Public. Routing through an IGW
-    // won't work because the ENIs don't get a Public IP.
-    // Why are we not simply forcing vpcSubnets? Because you might still be choosing
-    // Isolated networks or selecting among 2 sets of Private subnets by name.
     const { subnetIds } = props.vpc.selectSubnets(props.vpcSubnets);
-    const publicSubnetIds = new Set(props.vpc.publicSubnets.map(s => s.subnetId));
-    for (const subnetId of subnetIds) {
-      if (publicSubnetIds.has(subnetId)) {
-        throw new Error('Not possible to place Lambda Functions in a Public subnet');
```

Comment:
If a lambda was placed in a public subnet, then it cannot access the internet without additional set up on the VPC - https://stackoverflow.com/a/52994841/915163.
This is a common mistake new customers will make and not realize what the issue without further digging.

This was the original reason this restriction was put in place. Simply removing it may cause for customers to commit the same mistake. 

Can we instead add another property to `FunctionProps`, like `allowPublicSubnet`, that defaults to `false` and needs to be explicitly flipped to `true` to work around this error message? The error message can be expanded to document this limitation and instruct the user to set the flag if they are ok with this caveat.

Reply:
> If a lambda was placed in a public subnet, then it cannot access the internet without additional set up on the VPC - https://stackoverflow.com/a/52994841/915163.
> This is a common mistake new customers will make and not realize what the issue without further digging.
> 
> This was the original reason this restriction was put in place. Simply removing it may cause for customers to commit the same mistake.
> 
> Can we instead add another property to `FunctionProps`, like `allowPublicSubnet`, that defaults to `false` and needs to be explicitly flipped to `true` to work around this error message? The error message can be expanded to document this limitation and instruct the user to set the flag if they are ok with this caveat.

Yes that makes sense to me.